[CI] CachedBlobContainerIndexInputTests.testRandomReads failing

Saw 4 failures in the past two days.

**Build scan**:
https://gradle-enterprise.elastic.co/s/2hkq7vdyvz3zy/tests/:x-pack:plugin:searchable-snapshots:test/org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests/testRandomReads#1

**Repro line**:
`./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests.testRandomReads" -Dtests.seed=65316F60EB4EFBC7 -Dtests.security.manager=true -Dtests.locale=ar-JO -Dtests.timezone=Etc/GMT-0 -Druntime.java=8`

**Reproduces locally?**:
Did not reproduce locally

**Applicable branches**:
master and 7.x

**Failure history**:
https://gradle-enterprise.elastic.co/scans/tests?search.relativeStartTime=P7D&search.timeZoneId=America/Los_Angeles&tests.container=org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests&tests.disabledDistributions=WyJvdXRjb21lOnBhc3NlZCJd&tests.sortField=FAILED&tests.test=testRandomReads&tests.unstableOnly=true

**Failure excerpt**:
```
org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests > testRandomReads FAILED
    com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=47, name=elasticsearch[testRandomReads][generic][T#1], state=RUNNABLE, group=TGRP-CachedBlobContainerIndexInputTests]
        at __randomizedtesting.SeedInfo.seed([65316F60EB4EFBC7:5033CD9301E25FED]:0)

        Caused by:
        java.lang.AssertionError: java.lang.IllegalArgumentException: Failed to find a Lucene index for cache file path [/dev/shm/elastic+elasticsearch+7.x+periodic/x-pack/plugin/searchable-snapshots/build/testrun/test/temp/org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests_65316F60EB4EFBC7-001/tempDir-007/Jd_BIJGoTvmRmqAkG5fn5A]
            at __randomizedtesting.SeedInfo.seed([65316F60EB4EFBC7]:0)
            at org.elasticsearch.xpack.searchablesnapshots.cache.CacheService.synchronizeCache(CacheService.java:538)
            at org.elasticsearch.xpack.searchablesnapshots.cache.CacheService$CacheSynchronizationTask.runInternal(CacheService.java:575)
            at org.elasticsearch.common.util.concurrent.AbstractAsyncTask.run(AbstractAsyncTask.java:144)
            at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
            at java.lang.Thread.run(Thread.java:748)

            Caused by:
            java.lang.IllegalArgumentException: Failed to find a Lucene index for cache file path [/dev/shm/elastic+elasticsearch+7.x+periodic/x-pack/plugin/searchable-snapshots/build/testrun/test/temp/org.elasticsearch.index.store.cache.CachedBlobContainerIndexInputTests_65316F60EB4EFBC7-001/tempDir-007/Jd_BIJGoTvmRmqAkG5fn5A]
                at org.elasticsearch.xpack.searchablesnapshots.cache.PersistentCache.lambda$getWriter$1(PersistentCache.java:127)
                at java.util.Optional.orElseThrow(Optional.java:290)
                at org.elasticsearch.xpack.searchablesnapshots.cache.PersistentCache.getWriter(PersistentCache.java:127)
                at org.elasticsearch.xpack.searchablesnapshots.cache.PersistentCache.addCacheFile(PersistentCache.java:133)
                at org.elasticsearch.xpack.searchablesnapshots.cache.CacheService.synchronizeCache(CacheService.java:533)
                ... 6 more
```

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCache.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCache.java
@@ -64,6 +64,7 @@ import java.nio.file.Path;
 import java.nio.file.SimpleFileVisitor;
 import java.nio.file.attribute.BasicFileAttributes;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashMap;
@@ -124,7 +125,7 @@ public class PersistentCache implements Closeable {
             return writers.stream()
                 .filter(writer -> path.startsWith(writer.nodePath().path))
                 .findFirst()
-                .orElseThrow(() -> new IllegalArgumentException("Failed to find a Lucene index for cache file path [" + path + ']'));
+                .orElseThrow(() -> new PersistentCacheIndexNotFoundException(nodeEnvironment, cacheFile));
         }
     }
 
@@ -650,5 +651,20 @@ public class PersistentCache implements Closeable {
             Files.createDirectories(snapshotCacheRootDir);
         }
         return snapshotCacheRootDir;
+    }
+
+    /**
+     * Exception thrown when the {@link CacheIndexWriter} corresponding to a given {@link CacheFile} cannot be found.
+     */
+    static class PersistentCacheIndexNotFoundException extends IllegalArgumentException {
+
+        PersistentCacheIndexNotFoundException(NodeEnvironment nodeEnvironment, CacheFile cacheFile) {
+            super(
+                "Persistent cache index not found for cache file path ["
+                    + cacheFile.getFile()
+                    + "] using node paths "
+                    + Arrays.toString(nodeEnvironment.nodeDataPaths())
+            );
+        }
     }
 }

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java
@@ -64,7 +64,6 @@ import org.elasticsearch.common.util.MockBigArrays;
 import org.elasticsearch.common.xcontent.NamedXContentRegistry;
 import org.elasticsearch.core.internal.io.IOUtils;
 import org.elasticsearch.env.Environment;
-import org.elasticsearch.env.NodeEnvironment;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.seqno.SequenceNumbers;
@@ -93,7 +92,6 @@ import java.io.Closeable;
 import java.io.EOFException;
 import java.io.FileNotFoundException;
 import java.io.IOException;
-import java.io.UncheckedIOException;
 import java.nio.ByteBuffer;
 import java.nio.channels.SeekableByteChannel;
 import java.nio.charset.StandardCharsets;
@@ -578,14 +576,9 @@ public class SearchableSnapshotDirectoryTests extends AbstractSearchableSnapshot
                 final BlobContainer blobContainer = repository.shardContainer(indexId, shardId.id());
                 final BlobStoreIndexShardSnapshot snapshot = repository.loadShardSnapshot(blobContainer, snapshotId);
 
-                final Path shardDir;
-                try {
-                    shardDir = new NodeEnvironment.NodePath(createTempDir()).resolve(shardId);
-                } catch (IOException e) {
-                    throw new UncheckedIOException(e);
-                }
+                final Path shardDir = randomShardPath(shardId);
                 final ShardPath shardPath = new ShardPath(false, shardDir, shardDir, shardId);
-                final Path cacheDir = createTempDir();
+                final Path cacheDir = Files.createDirectories(resolveSnapshotCache(shardDir).resolve(snapshotId.getUUID()));
                 final CacheService cacheService = defaultCacheService();
                 releasables.add(cacheService);
                 cacheService.start();

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInputTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInputTests.java
@@ -13,7 +13,6 @@ import org.elasticsearch.common.blobstore.BlobContainer;
 import org.elasticsearch.common.blobstore.support.FilterBlobContainer;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.ByteSizeValue;
-import org.elasticsearch.env.NodeEnvironment;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.index.shard.ShardPath;
 import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot;
@@ -30,8 +29,8 @@ import java.io.EOFException;
 import java.io.FilterInputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.UncheckedIOException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.HashSet;
 import java.util.Objects;
@@ -44,6 +43,7 @@ import static org.elasticsearch.index.store.cache.TestUtils.singleBlobContainer;
 import static org.elasticsearch.index.store.cache.TestUtils.singleSplitBlobContainer;
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_CACHE_ENABLED_SETTING;
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_CACHE_PREWARM_ENABLED_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.cache.CacheService.resolveSnapshotCache;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThanOrEqualTo;
@@ -89,14 +89,9 @@ public class CachedBlobContainerIndexInputTests extends AbstractSearchableSnapsh
                 }
 
                 final boolean recoveryFinalizedDone = randomBoolean();
-                final Path shardDir;
-                try {
-                    shardDir = new NodeEnvironment.NodePath(createTempDir()).resolve(shardId);
-                } catch (IOException e) {
-                    throw new UncheckedIOException(e);
-                }
+                final Path shardDir = randomShardPath(shardId);
                 final ShardPath shardPath = new ShardPath(false, shardDir, shardDir, shardId);
-                final Path cacheDir = createTempDir();
+                final Path cacheDir = Files.createDirectories(resolveSnapshotCache(shardDir).resolve(snapshotId.getUUID()));
                 try (
                     SearchableSnapshotDirectory directory = new SearchableSnapshotDirectory(
                         () -> blobContainer,
@@ -190,14 +185,9 @@ public class CachedBlobContainerIndexInputTests extends AbstractSearchableSnapsh
             );
 
             final BlobContainer blobContainer = singleBlobContainer(blobName, input);
-            final Path shardDir;
-            try {
-                shardDir = new NodeEnvironment.NodePath(createTempDir()).resolve(shardId);
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
-            }
+            final Path shardDir = randomShardPath(shardId);
             final ShardPath shardPath = new ShardPath(false, shardDir, shardDir, shardId);
-            final Path cacheDir = createTempDir();
+            final Path cacheDir = Files.createDirectories(resolveSnapshotCache(shardDir).resolve(snapshotId.getUUID()));
             try (
                 SearchableSnapshotDirectory searchableSnapshotDirectory = new SearchableSnapshotDirectory(
                     () -> blobContainer,

